ksql-engine/src/main/java/io/confluent/ksql/topic/TopicDeleteInjector.java

```diff
@@ -0,0 +1,119 @@
+import com.google.common.collect.ImmutableList;
+import io.confluent.kafka.schemaregistry.client.SchemaRegistryClient;
+import io.confluent.ksql.KsqlExecutionContext;
+import io.confluent.ksql.metastore.MetaStore;
+import io.confluent.ksql.metastore.model.StructuredDataSource;
+import io.confluent.ksql.parser.SqlFormatter;
+import io.confluent.ksql.parser.tree.DropStatement;
+import io.confluent.ksql.parser.tree.Statement;
+import io.confluent.ksql.schema.registry.SchemaRegistryUtil;
+import io.confluent.ksql.serde.DataSource;
+import io.confluent.ksql.services.KafkaTopicClient;
+import io.confluent.ksql.statement.ConfiguredStatement;
+import io.confluent.ksql.statement.Injector;
+import io.confluent.ksql.util.ExecutorUtil;
+import io.confluent.ksql.util.ExecutorUtil.RetryBehaviour;
+import io.confluent.ksql.util.KsqlConstants;
+import io.confluent.ksql.util.KsqlException;
+import java.util.Objects;
+
+/**
+ * This {@code Injector} will delete the topic associated with a
+ * {@code DROP [STREAM/TABLE] X DELETE TOPIC}. Any other statements are just
+ * passed through. Furthermore, it will remove the DELETE TOPIC clause from
+ * the statement, indicating that the operation has already been done.
+ *
+ * <p>If the topic being deleted is {@link DataSource.DataSourceSerDe#AVRO},
+ * this injector will also clean up the corresponding schema in the schema
+ * registry.
+ */
+public class TopicDeleteInjector implements Injector {
+
+  private final MetaStore metastore;
+  private final KafkaTopicClient topicClient;
+  private final SchemaRegistryClient schemaRegistryClient;
+
+  public TopicDeleteInjector(final KsqlExecutionContext executionContext) {
+    this(
+        Objects.requireNonNull(executionContext, "executionContext").getMetaStore(),
+        executionContext.getServiceContext().getTopicClient(),
+        executionContext.getServiceContext().getSchemaRegistryClient()
+    );
+  }
+
+  @VisibleForTesting
+  TopicDeleteInjector(
+      final MetaStore metastore,
+      final KafkaTopicClient serviceContext,
+      final SchemaRegistryClient schemaRegistryClient
+  ) {
+    this.metastore = Objects.requireNonNull(metastore, "metastore");
+    this.topicClient = Objects.requireNonNull(serviceContext, "topicClient");
+    this.schemaRegistryClient =
+        Objects.requireNonNull(schemaRegistryClient, "schemaRegistryClient");
+  }
+
+  @SuppressWarnings("unchecked")
+  @Override
+  public <T extends Statement> ConfiguredStatement<T> inject(
+      final ConfiguredStatement<T> statement) {
+    if (!(statement.getStatement() instanceof DropStatement)) {
+      return statement;
+    }
+
+    final DropStatement dropStatement = (DropStatement) statement.getStatement();
+    if (!dropStatement.isDeleteTopic()) {
+      return statement;
+    }
+
+    final String sourceName = dropStatement.getName().getSuffix();
+    final StructuredDataSource<?> source = metastore.getSource(sourceName);
+
+    if (source != null) {
```

Comment:
I believe this code path is still hit if `DROP SOURCE IF EXISTS X`

Reply:
Can you elaborate? As I'm not sure I follow.... (or we can jump on a call).

What I'm saying is that before we reach this point in the code, (or maybe after), something is already checking (and potentially failing) is the source doesn't exist. So adding a second check and throw seems superfluous.

For example, for the rest-api, `KsqlResource` will already have validated the statement - which would fail if the source did not exist. Doesn't headless mode also validate statements? So this code can not be called with out the source existing.

I guess what I'm saying is that it's not the job of this bit of code to validate that the source exists. So it should not do so. 

On the flip side, a defensive programmer could argue (and in the next PR I likely will! ;) ), that we _should_ throw an exception _because_ we never expect this to be null.   In which case I'd suggest:
 1. Throwing an `IllegalStateException` as `KsqlException` is more about user facing exceptions and this is more about a programming error.
 2. Inverting the if statement.  It's a style thing, but I think:

```java
   if (badThingHappened) {
      throw new AppropriateException(usefulErrorMessage);
  }

  ... carry on with good path
```

Is easier to grok than:

```java
    if (badThingDidNotHappen) {
         then
         do
         lots
         of
         other
         stuff
         which
         can
         be
         many
         lines
   } else {
       throw new AppropriateException(usefulErrorMessage);
  }
```

Mainly because by the time I get to the `else` I've forgotten what the `if` was and have to scroll back up! ;)

Reply:
> What I'm saying is that before we reach this point in the code, (or maybe after), something is already checking (and potentially failing) is the source doesn't exist. So adding a second check and throw seems superfluous.

What I meant is that this isn't necessarily true. The validator runs the command, the command has the fully code: 
```
    if (dataSource == null) {
      if (ifExists) {
        return new DdlCommandResult(true, "Source " + sourceName + " does not exist.");
      }
      throw new KsqlException("Source " + sourceName + " does not exist.");
    }
```

There will be no issue if the dataSource is null, it'll go ahead and enqueue the command:
```
ksql> DROP STREAM DOESNTEXIST;
Source DOESNTEXIST does not exist.
ksql> DROP STREAM IF EXISTS DOESNTEXIST;

 Message
------------------------------------
 Source DOESNTEXIST does not exist.
------------------------------------
```
The first is a failure, the second is a successful command execution.

Reply:
Do we need to enqueue the command if the source does not exist? :D :D :D

Reply:
Probably not - but if we have an `IF EXISTS` semantics, then it isn't invalid. I think having special semantics for it would be more code burden that it's worth - we should either remove the `IF EXISTS` or treat it just like any other `DROP` command (i.e. enqueue it even if there's no source)